kong/clustering.lua

```diff
@@ -0,0 +1,415 @@
+      if not ok then
+        ngx_log(ngx_ERR, "unable to update in-memory cluster status: ", err)
+      end
+    end
+  end)
+
+  while not exiting() do
+    local ok, err = sem:wait(10)
+    if ok then
+      local config = table.remove(queue, 1)
+      assert(config, "config queue can not be empty after semaphore returns")
+
+      local _, err = wb:send_binary(cjson_encode({ type = "reconfigure",
+                                                       config_table = config,
+                                                     }))
+      if err then
+        ngx_log(ngx_ERR, "unable to send updated configuration to node: ", err)
+
+      else
+        ngx_log(ngx_DEBUG, "sent config update to node")
+      end
+
+    else -- not ok
+      if err ~= "timeout" then
+        ngx_log(ngx_ERR, "semaphore wait error: ", err)
+      end
+    end
+  end
+end
+
+
+function _M.get_status()
+  local result = new_tab(0, 8)
+
+  for _, n in ipairs(shdict:get_keys()) do
+    result[n] = cjson_decode(shdict:get(n))
+  end
+
+
+  return result
+end
+
+
+local function push_config(config_table)
+  local n = 0
+
+  for _, queue in pairs(clients) do
+    table.insert(queue, config_table)
+    queue.sem:post()
+
+    n = n + 1
+  end
+
+  ngx_log(ngx_DEBUG, "config pushed to ", n, " clients")
+end
+
+
+local function init_mtls(conf)
+  local f, err = io_open(conf.cluster_cert, "r")
+  if not f then
+    return nil, "unable to open cluster cert file: " .. err
+  end
+
+  local cert
+  cert, err = f:read("*a")
+  if not cert then
+    f:close()
+    return nil, "unable to read cluster cert file: " .. err
+  end
+
+  f:close()
+
+  CERT, err = ssl.parse_pem_cert(cert)
+  if not CERT then
+    return nil, err
+  end
+
+  cert = openssl_x509.new(cert, "PEM")
+
+  CERT_DIGEST = cert:digest("sha256")
+
+  f, err = io_open(conf.cluster_cert_key, "r")
+  if not f then
+    return nil, "unable to open cluster cert key file: " .. err
+  end
+
+  local key
+  key, err = f:read("*a")
+  if not key then
+    f:close()
+    return nil, "unable to read cluster cert key file: " .. err
+  end
+
+  f:close()
+
+  CERT_KEY, err = ssl.parse_pem_priv_key(key)
+  if not CERT_KEY then
+    return nil, err
+  end
+
+  return true
+end
+
+
+function _M.init(conf)
+  assert(conf, "conf can not be nil", 2)
+
+  if conf.role == "data_plane" or conf.role == "control_plane" then
+    assert(init_mtls(conf))
+  end
+end
+
+
+function _M.init_worker(conf)
+  assert(conf, "conf can not be nil", 2)
+
+  if conf.role == "data_plane" then
+    -- ROLE = "data_plane"
+
+    if ngx.worker.id() == 0 then
+      local f = io_open(CONFIG_CACHE, "r")
+      if f then
+        local config, err = f:read("*a")
+        if not config then
+          ngx_log(ngx_ERR, "unable to read cached config file: ", err)
+        end
+
+        f:close()
+
+        if config then
+          ngx_log(ngx_INFO, "found cached copy of data-plane config, loading..")
+          config = cjson_decode(config)
```

Comment:
Error handling in this call (as well as a in few of the other `cjson_decode` calls)?